components/style/rule_tree/mod.rs

```diff
@@ -580,13 +586,13 @@ impl StrongRuleNode {
                 let existing =
                     next_sibling_ptr.compare_and_swap(ptr::null_mut(),
                                                       new_ptr,
-                                                      Ordering::SeqCst);
+                                                      Ordering::Acquire);
```

Comment:
> Aren't you mixing reads of the fields on the nodes vs reads on the atomic pointers?

Unless I'm very confused here (which is a real possibility), I don't think so.

From my understanding, memory ordering isn't about ensuring any sort of consistency with regards to the actual atomic being loaded/stored (those operations are already atomic). Rather, memory ordering is about consistency in how other nearby, potentially-_non_-atomic operations are observed by various threads, in particularly how they may be reordered by the compiler and by the CPU. Sequential Consistency means that no other operations are observably reordered across the Sequentially-Consistent atomic operation, from the perspective of any thread. Release/Acquire means that no write operations performed by a thread manipulating field X with |Release| will be observed to be reordered across the change to X _from the perspective_ of a thread reading field X with |Acquire|. Other threads can observe entirely different global orderings, because they're not "synchronizing" on the same memory address. [1]

So in this case, I believe we need Release when CAS-ing in the new entries, because otherwise the initialization of the new entry may not be "flushed" before the memory is accessed by other threads. And those threads need to obtain the pointers with Release, so that they are guaranteed to view the "flushed" writes.

If I'm wrong here, please tell me, because I would be very interested to know that. :-)

[1] http://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering